### ISSUE TITLE
`Error: [omnidoc] No source found for path` after git commit while the dev server is running

### ISSUE THREAD

**Issue by noxify** [souporserious/omnidoc#174]
Hi,

Not sure if it's a bug or just user handling :) 

1. I made a change in a file inside my project ( e.g. `next.config.mjs` )
2. Saved the file => one uncommited change
3. Started the dev server
4. Opened a page => everything is fine so far and everything is working
    * http://localhost:3000/docs/build-your-application/routing/pages-and-layouts
6. Commited the change
7. Got the following error ( here the complete log )

```
❯ pnpm dev

> mdxts-docs-template@0.1.1 dev /Users/marcusreinhardt/Development/mdxts-docs-template
> concurrently "omnidoc watch" "next dev"

[1]   ▲ Next.js 14.2.7
[1]   - Local:        http://localhost:3000
[1] 
[1]  ✓ Starting...
[1]    automatically enabled Fast Refresh for 1 custom loader
[1]  ✓ Ready in 2.2s
[1]  ○ Compiling /docs/[...slug] ...
[1]  ✓ Compiled /docs/[...slug] in 3.6s (1382 modules)
[1]  ✓ Compiled in 47ms (355 modules)
[1] {
[1]   tree: [
[1]     {
[1]       title: 'Getting Started',
[1]       path: '/docs/getting-started',
[1]       slug: [Array],
[1]       order: '01',
[1]       depth: 1,
[1]       children: [Array]
[1]     },
[1]     {
[1]       title: 'Build Your Application',
[1]       path: '/docs/build-your-application',
[1]       slug: [Array],
[1]       order: '02.01',
[1]       depth: 1,
[1]       children: [Array]
[1]     }
[1]   ]
[1] }
[1]  GET /docs/build-your-application/routing/pages-and-layouts 200 in 7458ms
[0] node:fs:441
[0]     return binding.readFileUtf8(path, stringToFlags(options.flag));
[0]                    ^
[0] 
[0] Error: EISDIR: illegal operation on a directory, read
[0]     at Object.readFileSync (node:fs:441:20)
[0]     at NodeRuntimeFileSystem.readFileSync (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/@ts-morph+common@0.24.0/node_modules/@ts-morph/common/dist/ts-morph-common.js:1275:30)
[0]     at RealFileSystemHost.readFileSync (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/@ts-morph+common@0.24.0/node_modules/@ts-morph/common/dist/ts-morph-common.js:1931:23)
[0]     at TransactionalFileSystem.readFileSync (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/@ts-morph+common@0.24.0/node_modules/@ts-morph/common/dist/ts-morph-common.js:2514:33)
[0]     at TransactionalFileSystem.readFileIfExistsSync (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/@ts-morph+common@0.24.0/node_modules/@ts-morph/common/dist/ts-morph-common.js:2500:25)
[0]     at CompilerFactory.addOrGetSourceFileFromFilePath (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/ts-morph@23.0.0/node_modules/ts-morph/dist/ts-morph.js:19635:62)
[0]     at DirectoryCoordinator.addSourceFileAtPathIfExists (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/ts-morph@23.0.0/node_modules/ts-morph/dist/ts-morph.js:19103:38)
[0]     at DirectoryCoordinator.addSourceFileAtPath (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/ts-morph@23.0.0/node_modules/ts-morph/dist/ts-morph.js:19109:33)
[0]     at Project.addSourceFileAtPath (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/ts-morph@23.0.0/node_modules/ts-morph/dist/ts-morph.js:20574:51)
[0]     at FSWatcher.<anonymous> (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/omnidoc@3.0.0_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/omnidoc/dist/cjs/src/project/get-project.js:80:29) {
[0]   errno: -21,
[0]   code: 'EISDIR',
[0]   syscall: 'read'
[0] }
[0] 
[0] Node.js v22.5.1
[0] omnidoc watch exited with code 1
[1]  ✓ Compiled in 365ms (859 modules)
[1]  ⨯ Error: [omnidoc] No source found for path "/docs/build-your-application/routing/pages-and-layouts" at file pattern "@content/docs/**/*.{ts,tsx,mdx}":
[1] 
[1] You can fix this error by ensuring the following:
[1]   
[1]   - The ".omnidoc" directory was successfully created and your tsconfig.json file aliases "omnidoc/*" to ".omnidoc/*.ts" correctly.
[1]   - The file pattern is formatted correctly and targeting files that exist.
[1]   - Try refreshing the page or restarting server.
[1]   - If you continue to see this error, please file an issue: https://github.com/souporserious/omnidoc/issues
[1]     at DocsPage (./src/app/docs/[...slug]/page.tsx:61:53)
[1] digest: "1461207223"
[1] [omnidoc] Attempting to reconnect to WebSocket server... (1/5)
[1] Error: [omnidoc] WebSocket client error:  
[1] 
[1] This was most likely caused by the "omnidoc" server not running.
[1]     at #handleError (webpack-internal:///(rsc)/./node_modules/.pnpm/omnidoc@3.0.0_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/omnidoc/dist/esm/project/rpc/client.js:53:15)
[1]     at callListener (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:290:14)
[1]     at WebSocket.onError (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:230:9)
[1]     at WebSocket.emit (node:events:520:28)
[1]     at emitErrorAndClose (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:1041:13)
[1]     at ClientRequest.eval (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:881:5)
[1]     at ClientRequest.emit (node:events:520:28)
[1]     at Socket.socketErrorListener (node:_http_client:495:9)
[1]     at Socket.emit (node:events:520:28)
[1]     at emitErrorNT (node:internal/streams/destroy:170:8)
[1]     at emitErrorCloseNT (node:internal/streams/destroy:129:3)
[1]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
[1]  ⨯ uncaughtException: Error: [omnidoc] WebSocket client error:  
[1] 
[1] This was most likely caused by the "omnidoc" server not running.
[1]     at #handleError (webpack-internal:///(rsc)/./node_modules/.pnpm/omnidoc@3.0.0_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/omnidoc/dist/esm/project/rpc/client.js:53:15)
[1]     at callListener (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:290:14)
[1]     at WebSocket.onError (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:230:9)
[1]     at WebSocket.emit (node:events:520:28)
[1]     at emitErrorAndClose (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:1041:13)
[1]     at ClientRequest.eval (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:881:5)
[1]     at ClientRequest.emit (node:events:520:28)
[1]     at Socket.socketErrorListener (node:_http_client:495:9)
[1]     at Socket.emit (node:events:520:28)
[1]     at emitErrorNT (node:internal/streams/destroy:170:8)
[1]     at emitErrorCloseNT (node:internal/streams/destroy:129:3)
[1]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
[1]  ⨯ uncaughtException: Error: [omnidoc] WebSocket client error:  
[1] 
[1] This was most likely caused by the "omnidoc" server not running.
[1]     at #handleError (webpack-internal:///(rsc)/./node_modules/.pnpm/omnidoc@3.0.0_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/omnidoc/dist/esm/project/rpc/client.js:53:15)
[1]     at callListener (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:290:14)
[1]     at WebSocket.onError (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/event-target.js:230:9)
[1]     at WebSocket.emit (node:events:520:28)
[1]     at emitErrorAndClose (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:1041:13)
[1]     at ClientRequest.eval (webpack-internal:///(rsc)/./node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket.js:881:5)
[1]     at ClientRequest.emit (node:events:520:28)
[1]     at Socket.socketErrorListener (node:_http_client:495:9)
[1]     at Socket.emit (node:events:520:28)
[1]     at emitErrorNT (node:internal/streams/destroy:170:8)
[1]     at emitErrorCloseNT (node:internal/streams/destroy:129:3)
[1]     at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
```

To reproduce the issue you can use this repo: https://github.com/noxify/mdxts-docs-template

My setup is the following:

```
> next info


Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6030
  Available memory (MB): 36864
  Available CPU cores: 12
Binaries:
  Node: 22.5.1
  npm: 10.8.2
  Yarn: N/A
  pnpm: 9.9.0
Relevant Packages:
  next: 14.2.7 // Latest available version is detected (14.2.7).
  eslint-config-next: 14.2.7
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.5.4
Next.js Config:
  output: export
```

**Comment by souporserious** [comment]
Hm it looks like `ts-morph` is getting tripped up somehow 🤔 assuming it's where the files are synced in the watcher. I'll take a look and see if I can reproduce and work on a fix!

**Comment by noxify** [comment]
If you need help with the reproduction/if it helps, we could have a call next week, then I can share my screen and we can debug it together.

Will check it next week with one of the existing example (called from the omnidoc repo), maybe I can reproduce it, too.

Please let me know if you need some additional information or something else ✌️

**Comment by souporserious** [comment]
Sounds good! I'd be happy to hop on a call. I'll let you know if I can't figure it out and we can schedule something for early next week.

**Comment by noxify** [comment]
Good morning @souporserious,

Is it possible that the omnidoc watch modes checks all directories and files in the project including the hidden ones?

What I mean? Maybe there is a change inside the `.git` directory, triggered by the `git commit` command - and the ts-morph watchs also on this directory which could produce the error?

Just an idea which I had after I woked up 🙈

**Comment by noxify** [comment]
Maybe adding the `.git` directory to the default ignore would help.

https://github.com/souporserious/omnidoc/blob/main/packages/omnidoc/src/project/server.ts

But will try it also with `omnidoc next dev` - maybe there is a different behavior 🤷‍♂️

**Comment by noxify** [comment]
With `omnidoc next dev` I get the following error:

```
 ✓ Compiled in 40ms (563 modules)
 ⨯ Error: [omnidoc] No source found for path "/docs/build-your-application/routing/pages-and-layouts" at file pattern "@content/docs/**/*.{ts,tsx,mdx}":

You can fix this error by ensuring the following:
  
  - The ".omnidoc" directory was successfully created and your tsconfig.json file aliases "omnidoc/*" to ".omnidoc/*.ts" correctly.
  - The file pattern is formatted correctly and targeting files that exist.
  - Try refreshing the page or restarting server.
  - If you continue to see this error, please file an issue: https://github.com/souporserious/omnidoc/issues
    at DocsPage (./src/app/docs/[...slug]/page.tsx:61:53)
digest: "1461207223"
 ✓ Compiled in 879ms (1402 modules)
```

** but** the process isn't killed.

**Comment by noxify** [comment]
When the process dies with `omnidoc next dev` and I kill the process afterwards to restart it, I mentioned that the subprocess is still "running" --> Port `3000` is in use and it starts the next server with port `3001`.

**Comment by souporserious** [comment]
Awesome, glad the latest release fixed the sub process not cleaning up correctly! Also added `.git` to the default ignore list so hopefully that's fixed as well 🤞. Appreciate all of the feedback! I've only tested in monorepos so far so this really helps cover simple repos 🙏

**Comment by noxify** [comment]
> Awesome, glad the latest release fixed the sub process not cleaning up correctly!

Haven't tested it 🙈 but based
On the code changes, I assumed it's fixed, because you're always closing the process.

Do we have a pre-release with the current changes?

**Comment by noxify** [comment]
@souporserious tried to install the latest version, but I got the following error:

```
 ERR_PNPM_SPEC_NOT_SUPPORTED_BY_ANY_RESOLVER  restyle@catalog: isn't supported by any available resolver.

This error happened while installing the dependencies of omnidoc@3.1.0

An external package outside of the pnpm workspace declared a dependency using
the catalog protocol. This is likely a bug in that external package. Only
packages within the pnpm workspace may use catalogs. Usages of the catalog
protocol are replaced with real specifiers on 'pnpm publish'.

This is likely a bug in the publishing automation of this package. Consider filing
a bug with the authors of:

  omnidoc@3.1.0
```

System info is the same as in the initial post.
Updated the repo with the latest changes: https://github.com/noxify/mdxts-docs-template

**Comment by souporserious** [comment]
Whoops, sorry about that! Updated to the new [PNPM catalogs](https://pnpm.io/catalogs) feature and looks like the version I was using had a bug with `pnpm publish`. Fixed in `3.1.1`!

**Comment by noxify** [comment]
Good morning,

thanks for the fix - tested it quickly with release 3.1.1 and can confirm that both bugs seems to be fixed.

Tested the "git commit" issue ( `No source found for path...` ) with `omnidoc next dev` and `concurrently \"omnidoc watch\" \"next dev\"`.

Killing the process while running `omnidoc next dev` seems to kill the subprocess ( checked via `lsof -i:3000` ).

Thanks again for the fast responses and providing the fixes <3
